spec/controllers/work_groups_controller_spec.rb

```diff
@@ -1,17 +1,119 @@
 require 'rails_helper'
 
 RSpec.describe WorkGroupsController, type: :controller do
-  describe "GET #new" do
-    it "returns http success" do
+
+  let(:teacher)     { create(:teacher) }
+  let(:student1)    { create(:student) }
+  let(:student2)    { create(:student) }
+  let(:classroom)   { create(:classroom, user: teacher) }
+  let(:work_group1) { create(:work_group, classroom: classroom) }
+  let(:worksheet_template) { create(:worksheet_template, user: teacher) }
+  # let(:worksheet_template2) { create(:worksheet_template, image_url: 'google.com', user: teacher) }
+  let(:worksheet1) { create(:worksheet, work_group: work_group1, worksheet_template: worksheet_template) }
+
+  before do
+    request.headers['X-User-Email'] = teacher.email
+    request.headers['X-User-Token'] = teacher.authentication_token
+  end
+
+  describe 'GET #new' do
+    context 'success' do
+      it 'assigns correct new work group with params' do
+        worksheet1
+        put :new, params: {
+          work_group: { turn_time: 300_000 },
+          no_model_fields: {
+            emails: 'test1@gmail.com test2@gmail.com',
+            worksheet_url: 'https://res.cloudinary.com/naokimi/image/upload/v1563422680/p7ojmgdtwshkrhxmjzh1.jpg',
+            start_time: DateTime.now + 1.day,
+            start_date: Date.today
+          }
+        }
+        expect(controller.instance_variable_get(:@work_group).turn_time).to eq(300_000)
+        expect(WorkGroup.find_by_turn_time(300_000)).to be_instance_of(WorkGroup)
+        expect(controller.instance_variable_get(:@worksheet_urls)).to eq(['https://res.cloudinary.com/naokimi/image/upload/v1563422680/p7ojmgdtwshkrhxmjzh1.jpg'])
+        expect(response).to have_http_status(:success)
+      end
+
+      it 'assigns correct new work group given no params' do
+        worksheet1
+        get :new
+        # expect(controller.instance_variable_get(:@work_group).turn_time).to eq(nil)
+        expect(controller.instance_variable_get(:@worksheet_urls)).to eq(['https://res.cloudinary.com/naokimi/image/upload/v1563422680/p7ojmgdtwshkrhxmjzh1.jpg'])
+        expect(response).to have_http_status(:success)
+      end
+    end
+
+    it 'Worksheet has wrong url and returns empty' do
+      # worksheet_template2
       get :new
+      expect(controller.instance_variable_get(:@worksheet_urls)).to eq([])
       expect(response).to have_http_status(:success)
     end
-  end
 
-  describe "GET #create" do
-    it "returns http success" do
-      get :create
-      expect(response).to have_http_status(:success)
+  describe 'GET #create' do
+    context 'success' do
```

Comment:
an important pointing about writing unit tests is the concept of stubbing and mocking https://relishapp.com/rspec/rspec-mocks/v/3-9/docs
when you test a unit (i.e. a file), you want to test only the code that is inside that unit, and not other services that might be called. Those other services will have their own tests, so you should stub the calls and return values as if they were successful.

For example in this test you should be adding this block:
```
let(:work_group_demo_prepper) { double(WorkGroupDemoPrepper) }

context 'xxx' do
  before do
    allow(WorkGroupDemoPrepper).to receive(:new).and_return(work_group_demo_prepper)
    allow(work_group_demo_prepper).to receive(:call).and_return(xxx)
  end

  it 'xxx'
end
```

If you don't implement stubbing and mocking, every time you run tests you'll be making API calls to Cloudinary (uploading files), which can end up costing me money 😅